jobs/MeterInstanceJob.js

```diff
@@ -0,0 +1,186 @@
+    if (_.indexOf(excluded_service_names, serviceName) >= 0) {
+      const excluded_plans = _
+        .chain(config.metering.excluded_service_plans)
+        .filter(p => p.service_name === serviceName)
+        .head()
+        .get('plan_sku_names')
+        .value();
+      logger.info(`Excluded plans for ${serviceName}:`, excluded_plans);
+      if (_.indexOf(excluded_plans, planName) >= 0) {
+        logger.info(`The metering event guid: ${options.id} with ${serviceName}, ${planName} is excluded`);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static enrichEvent(options) {
+    // Add region , service name and plan sku name of the event
+    options = JSON.parse(options);
+    const serviceId = _.get(options, 'service.id');
+    const planId = _.get(options, 'service.plan');
+    logger.info(`Enriching the metering event ${serviceId}, plan: ${planId}`);
+    options.service.id = this.getServiceNameFromServiceGUID(serviceId);
+    options.service.plan = this.getPlanSKUFromPlanGUID(serviceId, planId);
+    options.consumer.region = config.metering.region;
+    return options;
+  }
+
+  static getServiceNameFromServiceGUID(serviceGuid) {
+    return _.chain(catalog.toJSON().services)
+      .map((s) => s.id === serviceGuid ? s.name : undefined)
+      .filter(s => s !== undefined)
+      .head()
+      .value();
+  }
+
+  static getPlanSKUFromPlanGUID(serviceGuid, planGuid) {
+    logger.info(`Getting Plan SKU for service ${serviceGuid}, plan: ${planGuid}`);
+    const service = _.chain(catalog.toJSON().services)
+      .map((s) => s.id === serviceGuid ? s : undefined)
+      .filter(s => s !== undefined)
+      .head()
+      .value();
+    return _
+      .chain(service.plans)
+      .map((p) => p.id === planGuid ? p.sku_name : undefined)
+      .filter(p => p !== undefined)
+      .head()
+      .value();
+  }
+
+  static getPlanNameFromPlanGUID(serviceGuid, planGuid) {
+    const service = _.chain(catalog.toJSON().services)
+      .map((s) => s.id === serviceGuid ? s : undefined)
+      .filter(s => s !== undefined)
+      .head()
+      .value();
+    return _
+      .chain(service.plans)
+      .map((p) => p.id === planGuid ? p.name : undefined)
+      .filter(p => p !== undefined)
+      .head()
+      .value();
+  }
+
+  static sendEvent(event) {
+    logger.info('Sending event:', event);
+    if (this.isServicePlanExcluded(event.spec.options)) {
+      return Promise.try(() => this.updateMeterState(CONST.METER_STATE.EXCLUDED, event))
+        .return(true);
+    }
+    return Promise
+      .try(() => this.enrichEvent(event.spec.options))
+      .then(enriched_usage_doc => {
+        logger.debug('Sending document:', enriched_usage_doc);
+        return maas.client.putUsageRecord({
+          usage: [enriched_usage_doc]
+        });
+      })
+      .then(validEvent => validEvent ? this.updateMeterState(CONST.METER_STATE.METERED, event) : false)
+      .return(true)
+      .catch(err => {
```

Comment:
What if `maas.client.putUsageRecord` worked but `this.updateMeterState` failed. Then ? :)

Reply:
Currently the job would retry the same document again in the next run.
Since we have unique ids for each metering doc it will simply result in updating the same entry.

Do we need to add further error handling here ?

Reply:
if retry is not a harm, then it is ok. Can you put a comment here about this discussion?